File: ios/Classes/Serializer/BannerInitCustomizationSerializer.swift
import Usercentrics

private func kotlinInt(_ value: Any?) -> KotlinInt? {
    guard let value = value as? Int else { return nil }
    return KotlinInt(int: Int32(value))
}

private func kotlinFloat(_ value: Any?) -> KotlinFloat? {
    guard let value = value as? Double else { return nil }
    return KotlinFloat(float: Float(value))
}

private func kotlinBool(_ value: Any?) -> KotlinBoolean? {
    guard let value = value as? Bool else { return nil }
    return KotlinBoolean(bool: value)
}

extension BannerInitCustomization {
    static func from(value: Any?) -> BannerInitCustomization? {
        guard let dict = value as? Dictionary<String, Any> else { return nil }

        var purposeListStyle: PurposeListStyle?
        if let purposeListStyleString = dict["purposeListStyle"] as? String {
            purposeListStyle = PurposeListStyle.initialize(from: purposeListStyleString)
        }

        return BannerInitCustomization(
            paddingTop: kotlinInt(dict["paddingTop"]),
            paddingBottom: kotlinInt(dict["paddingBottom"]),
            paddingStart: kotlinInt(dict["paddingStart"]),
            paddingEnd: kotlinInt(dict["paddingEnd"]),
            lineSpacingMultiplier: kotlinFloat(dict["lineSpacingMultiplier"]),
            titleFontSize: kotlinInt(dict["titleFontSize"]),
            bodyFontSize: kotlinInt(dict["bodyFontSize"]),
            linkFontSize: kotlinInt(dict["linkFontSize"]),
            titleFontBold: kotlinBool(dict["titleFontBold"]),
            headerPaddingTop: kotlinInt(dict["headerPaddingTop"]),
            headerPaddingSides: kotlinInt(dict["headerPaddingSides"]),
            headerPaddingBetweenElements: kotlinInt(dict["headerPaddingBetweenElements"]),
            buttonBorderColor: dict["buttonBorderColor"] as? String,
            buttonBorderWidth: kotlinInt(dict["buttonBorderWidth"]),
            purposeListStyle: purposeListStyle,
            stickyHeader: kotlinBool(dict["stickyHeader"]),
            hideLanguageSwitcher: kotlinBool(dict["hideLanguageSwitcher"]),
            buttonHeightDp: kotlinInt(dict["buttonHeightDp"]),
            buttonHorizontalPaddingDp: kotlinInt(dict["buttonHorizontalPaddingDp"]),
            buttonSpacingDp: kotlinInt(dict["buttonSpacingDp"]),
            linkUnderline: kotlinBool(dict["linkUnderline"]),
            showSecondLayerCloseButton: kotlinBool(dict["showSecondLayerCloseButton"]),
            tabFontSize: kotlinInt(dict["tabFontSize"]),
            tabActiveColor: dict["tabActiveColor"] as? String,
            denyAllButtonBackground: dict["denyAllButtonBackground"] as? String,
            acceptAllButtonBackground: dict["acceptAllButtonBackground"] as? String,
            linkColor: dict["linkColor"] as? String
        )
    }
}

extension PurposeListStyle {
    static func initialize(from value: String) -> PurposeListStyle? {
        switch value {
        case "BOXED":
            return PurposeListStyle.boxed
        case "FLAT":
            return PurposeListStyle.flat
        default:
            assert(false)
            return nil
        }
    }
}

File: android/src/main/kotlin/com/usercentrics/sdk/flutter/serializer/BannerInitCustomizationSerializer.kt
@file:Suppress("DEPRECATION")

package com.usercentrics.sdk.flutter.serializer

import com.usercentrics.sdk.BannerInitCustomization
import com.usercentrics.sdk.PurposeListStyle

internal fun Any?.deserializeBannerInitCustomization(): BannerInitCustomization? {
    if (this !is Map<*, *>) {
        return null
    }
    return BannerInitCustomization(
        paddingTop = this["paddingTop"] as? Int,
        paddingBottom = this["paddingBottom"] as? Int,
        paddingStart = this["paddingStart"] as? Int,
        paddingEnd = this["paddingEnd"] as? Int,
        lineSpacingMultiplier = (this["lineSpacingMultiplier"] as? Double)?.toFloat(),
        titleFontSize = this["titleFontSize"] as? Int,
        bodyFontSize = this["bodyFontSize"] as? Int,
        linkFontSize = this["linkFontSize"] as? Int,
        titleFontBold = this["titleFontBold"] as? Boolean,
        headerPaddingTop = this["headerPaddingTop"] as? Int,
        headerPaddingSides = this["headerPaddingSides"] as? Int,
        headerPaddingBetweenElements = this["headerPaddingBetweenElements"] as? Int,
        buttonBorderColor = this["buttonBorderColor"] as? String,
        buttonBorderWidth = this["buttonBorderWidth"] as? Int,
        purposeListStyle = (this["purposeListStyle"] as? String)?.let {
            PurposeListStyle.valueOf(it)
        },
        stickyHeader = this["stickyHeader"] as? Boolean,
        hideLanguageSwitcher = this["hideLanguageSwitcher"] as? Boolean,
        buttonHeightDp = this["buttonHeightDp"] as? Int,
        buttonHorizontalPaddingDp = this["buttonHorizontalPaddingDp"] as? Int,
        buttonSpacingDp = this["buttonSpacingDp"] as? Int,
        linkUnderline = this["linkUnderline"] as? Boolean,
        showSecondLayerCloseButton = this["showSecondLayerCloseButton"] as? Boolean,
        tabFontSize = this["tabFontSize"] as? Int,
        tabActiveColor = this["tabActiveColor"] as? String,
        denyAllButtonBackground = this["denyAllButtonBackground"] as? String,
        acceptAllButtonBackground = this["acceptAllButtonBackground"] as? String,
        linkColor = this["linkColor"] as? String,
    )
}
